csrc/preseg_passes/move_split_cat.cpp

```diff
@@ -105,6 +104,87 @@ std::pair<std::vector<PadOp*>, int64_t> getCatInputsAndAxis(CatOp* cat) {
   return {pads, cat_axis};
 }
 
+// If `exprs` are `SliceOp`s that form a split, returns the base tensor of the
+// split. Returns null otherwise.
+TensorView* exprsFormSplit(
+    const std::vector<Expr*>& exprs,
+    const int64_t split_axis) {
+  // Checks that all exprs are slices and are based on the
+  // same tensor. Otherwise, they don't form a split.
+  TensorView* split_in = nullptr;
+  for (Expr* e : exprs) {
+    auto* slice = dynamic_cast<SliceOp*>(e);
+    if (slice == nullptr) {
+      return nullptr;
+    }
+
+    if (split_in == nullptr) {
+      split_in = slice->in();
+    } else if (split_in != slice->in()) {
+      return nullptr;
+    }
+  }
+  NVF_ERROR(split_in != nullptr);
+
+  // Check that `exprs` (already known to be `SliceOp`s) form a split along
+  // `split_axis`.
+  //
+  // `split_ranges[i]` is the slice range of `exprs[i]` for the split axis.
+  std::vector<Slice> split_ranges;
+  split_ranges.reserve(exprs.size());
+  for (auto i : c10::irange(exprs.size())) {
+    auto* slice = exprs[i]->as<SliceOp>();
+    const std::vector<Slice>& slice_ranges = slice->getRanges();
+    // Check the steps are all one.
+    if (std::any_of(
+            slice_ranges.begin(),
+            slice_ranges.end(),
+            [](const Slice& slice_range) {
+              return !slice_range.step->isOne();
+            })) {
+      return nullptr;
+    }
+
+    // Check only the split axis is sliced.
+    for (auto j : c10::irange(
+             static_cast<int64_t>(slice->out()->getRootDomain().size()))) {
+      const bool sliced =
+          (slice->out()->getRootDomain()[j] !=
+           slice->out()->getMaybeRFactorDomain()[j]);
+      if ((j == split_axis) != sliced) {
+        return nullptr;
+      }
+    }
+
+    // Collect the slice range for the split axis.
+    split_ranges.push_back(slice_ranges[split_axis]);
+  }
+
+  if (!split_ranges.front().start->isZero()) {
```

Comment:
IIUC, the implicit assumption here is that the slice expression ranges are in sorted order?

Reply:
Is it possible for the slice expressions to not be in the order of their slice ranges? For e.g. if the order of slice expressions is inverted`s0 = in[:, 5:], s1 = in[:, :5]`

Reply:
> Yes, it's possible. What's the problem with that? Do you worry about that the optimization may misfire or may miss opportunities? 

That it may miss opportunities, although seems less probable.

> FWIW, `exprs` in `exprsFormSplit` are in the order they are concatenated.

Got it.